### ISSUE TITLE
Race conditions in inventory tracking, order, payment status...

### ISSUE THREAD

**Issue by umpirsky** [Sylius/Sylius#2776]
We are getting inventory deducted at exact same second, where one of them should have already been sold out. Looks like availability is checked at the same time, saying item is in stock, and then deducted twice by two different orders/users.

Similar thing happens with order status being changed to "confirmed" and "pending" at the same time, and then one override another. We sometimes see confirmed order with "pending" status, according to the change log they was changed at the same second.

@pjedrzejewski Are you aware of this issue? Can you suggest ways to fix it, I am interested to contribute it to Sylius. I was even thinking of MySQL row locking. :)

**Comment by pjedrzejewski** [comment]
Are we checking the inventory availability just before deducting the inventory? I remember that was causing some problems. We do not check inventory at the last step.

**Comment by umpirsky** [comment]
@pjedrzejewski First of all, thanks for blazing fast response, really appreciate that!

We are calling `[@sylius.order_processing.inventory_handler, 'updateInventory']` on `confirm`, and it should:

```php
if ($quantity < 1) {
    throw new \InvalidArgumentException('Quantity of units must be greater than 0.');
}
```

But looks like it happens that it is called at the same time when there are lots of users (race conditions), before decrease is actually flushed to db.

See you this weekend. ;)

**Comment by pjedrzejewski** [comment]
How often you run into this issue?

PS. Yes, see you soon, that is awesome! :)

**Comment by umpirsky** [comment]
@kayue More often with order / payment status, not sure how often, but it happens. Will find out how often exactly...

**Comment by kayue** [comment]
@pjedrzejewski @umpirsky Yes. I would say there is about 2 - 5% chance this might happen. (I have about 100 orders per day)

**Comment by patrick-mcdougle** [comment]
I think we can solve this with doctrine optimistic locking using quantity as the "version". Or pessimistic locking, wrapping the whole finalize order step in a transaction.

**Comment by michalmarcinkowski** [comment]
@patrick-mcdougle "Version" should be a separate field. Doctrine takes care of checking the Version before updating and updates the Version after each update to the record.

Using pessimistic locking, we have to be careful, because there is possibility of deadlock to occur.

**Comment by patrick-mcdougle** [comment]
If you're only locking one row per transaction (the variant), I don't think deadlocking is an issue. The "other" thread will just block until it can acquire the lock on that product variant. Unless there's something I'm not considering.

**Comment by umpirsky** [comment]
@michalmarcinkowski If I got the idea correctly, we should lock variant record before we check if item is in stock on final step in the checkout, and make sure version didn't change until order is finalized (this happens during one http request iirc)? If it is we handle `OptimisticLockException` and show some nice message to user, like "Woah, looks like someone just bought this item in front of your nose". :)

**Comment by michalmarcinkowski** [comment]
Yes, we should lock variant to be sure that nobody modified it in a meantime. If the variant will be modified we will get `OptimisticLockException`. It means that someone made a change to the object, but it doesn't mean the quantity is too low, so then we should refresh the variant (get the newest version from the database) and repeat the quantity check. If the quantity is too low, display message, when it's ok, try to update the variant.

I see a few things which will require investigation.

1. We operate on order, but we should lock only related variants.
2. We modify several variants, handing refresh will not be, so easy.
3. Handle case when some variants will be available, but some of them not.

Are you willing to work on that?

**Comment by umpirsky** [comment]
@michalmarcinkowski Yes, I am aware that change does not mean that item is sold out, but another thing we should be aware of is that concurrency can occur during this repeated check as well (immediately after). So, you refresh, and variant it is deducted by another user while you are checking, hm?

Not sure when I will be available to work on this, will get back to you and let you know if I do.

**Comment by michalmarcinkowski** [comment]
@umpirsky exactly, that's why we should repeat the quantity check together with locking variants again. We should have a loop, refresh and try to decrease, until `OptimisticLockException` is not thrown. Maybe we should also add condition to limit number of refreshes to some reasonable number to avoid getting stalled.

**Comment by umpirsky** [comment]
OK, sounds like we have a plan. Only a matter of time. :)

**Comment by michalmarcinkowski** [comment]
:+1:

**Comment by patrick-mcdougle** [comment]
Pessimistic locking would prevent the repeated concurrency problem outlined by @umpirsky If you hold a pessimistic lock, you know for sure that nobody is messing with that row because the db won't allow it.

**Comment by patrick-mcdougle** [comment]
And attempts to acquire the pessimistic lock will just block until the lock can be acquired (saving us the work of writing a loop)

**Comment by patrick-mcdougle** [comment]
This same concurrency problem exists in promotion/coupon counting as well, but it's less mission critical. It's not tragic if you over sell a coupon.

**Comment by michalmarcinkowski** [comment]
When using pessimistic locking, when we acquire the lock, nobody else can. The second problem is "However for Pessimistic Locking to work you have to disable the Auto-Commit Mode of your Database and start a transaction around your pessimistic lock" what can be an issue.

I don't want to suggest anything, IMO we should check this 2 strategies and decide which will be a better approach. I'm willing to investigate both of them when I will find some time (probably later than sooner :disappointed:).

**Comment by patrick-mcdougle** [comment]
Here's a little gist I threw together: https://gist.github.com/patrick-mcdougle/e2d7aa4d93b44b8247c5.

If you throw this action into some controller you can fire off a request to it with the sleep parameter set to a number of seconds. Then fire another request to the same endpoint (0 seconds of sleep) in another tab (incognito preferred because of how symfony locks the user session). Might want to test with other db vendors. I know this works with MySQL.

**Comment by patrick-mcdougle** [comment]
Basically we could use the event system to have a listener that locks and unlocks the database and we can set the priorities such that they wrap the code that needs to be inside it.

**Comment by patrick-mcdougle** [comment]
Well, we launched our site after being willing to accept the concurrency risk and it's worse than we thought. Over a period of a few days, our inventory count had drifted by ~500 units. It looks like I'll be cobbling something together for this sooner than later. I'll let you know what we did to fix this issue.

**Comment by pjedrzejewski** [comment]
@patrick-mcdougle Ouch, that sounds bad. Are you using vanilla checkout process or you did some modifications? We definitely need to take care of this soon.

**Comment by kayue** [comment]
@patrick-mcdougle that is a lot, you must have a lot of transaction.
Do you have problem with over-sold or?

**Comment by patrick-mcdougle** [comment]
Luckily we caught it before tragedy. Our current bandaid is to update inventory daily. I have a patch that I'm going to share soon, but I don't think it's an ideal one.

We're using a modified checkout process, but I think our solution will be a drop-in to the vanilla core or inventory bundle. It basically uses the inventory pre/post events to do what's outlined in my gist using pessimistic locking.

**Comment by patrick-mcdougle** [comment]
I've edited my gist from above with the listener we're using in production right now. This is on 0.14 of sylius, not sure what has changed since then, so don't know if this is a drop in for core now, but it's working for us. The challenge we're dealing with now is that some threads (in high concurrency) end up timing out waiting for the lock. This is particularly tragic when we've already charged the customer and the database transaction is rolled back without refunding the customer. But that's a different problem. Ideally, our checkout flow would retry the transaction or checkout should only auth the customer and some background process would actually capture the transaction, but that's a problem for a different day.

**Comment by psihius** [comment]
I have a suggestion that I used when dealing with money and balances.
In my experience dealing with money for 4 years is that whatever locking mechanic you try to use, it goes wrong anyway. The moment you have more than one web server, your only source to do locking is the database. Some have good tools for that, some not so. Some can lock a row, others can lock a table. In any case, I had dealt with too many deadlock situations in MySQL due - probably when PHP scripts died due to fatal error or some other problem and your database left with the lock active.
What I ended up doing, is using double read strategy in my code. Start a transaction, check the values, do the operation, re-read the value and check that it didn't go below 0. If everything is okay - commit. If not - rollback the whole thing and throw an error. This way you are guaranteed to catch the moment you go into negative.
Good thing a saw this, because if I launch the shop I develop with this not fixed, 500 unit drift from the above comment would be a mild case compared to my problems :(
